nameresolution/consul/consul.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	if len(services) == 0 {
+		return "", fmt.Errorf("no healthy services found with AppID:%s", req.ID)
+	}
+
+	shuffle := func(services []*consul.ServiceEntry) []*consul.ServiceEntry {
+		for i := len(services) - 1; i > 0; i-- {
+			rndbig, _ := rand.Int(rand.Reader, big.NewInt(int64(i+1)))
+			j := rndbig.Int64()
+
+			services[i], services[j] = services[j], services[i]
+		}
+
+		return services
+	}
+
+	svc := shuffle(services)[0]
+
+	// TODO - kubernetes behaviour?
+
+	addr := ""
+
+	if port, ok := svc.Service.Meta[cfg.DaprPortMetaKey]; ok {
+		if svc.Service.Address != "" {
+			addr = fmt.Sprintf("%s:%s", svc.Service.Address, port)
+		} else if svc.Node.Address != "" {
+			addr = fmt.Sprintf("%s:%s", svc.Node.Address, port)
+		} else {
+			return "", fmt.Errorf("no healthy services found with AppID:%s", req.ID)
+		}
+	} else {
+		return "", fmt.Errorf("target service AppID:%s found but DAPR_PORT missing from meta", req.ID)
+	}
+
+	return addr, nil
+}
+
+// getConfig configuration from metadata, defaults are best suited for self-hosted mode
+func getConfig(metadata nr.Metadata) (resolverConfig, error) {
+	var daprPort string
+	var ok bool
+	var err error
+	resolverCfg := resolverConfig{}
+
+	props := metadata.Properties
+
+	if daprPort, ok = props[nr.DaprPort]; !ok {
+		return resolverCfg, fmt.Errorf("nr metadata property missing: %s", nr.DaprPort)
+	}
+
+	cfg, err := parseConfig(metadata.Configuration)
+	if err != nil {
+		return resolverCfg, err
+	}
+
+	// set DaprPortMetaKey used for registring DaprPort and resolving from Consul
+	if cfg.DaprPortMetaKey == "" {
+		resolverCfg.DaprPortMetaKey = daprMeta
+	} else {
+		resolverCfg.DaprPortMetaKey = cfg.DaprPortMetaKey
+	}
+
+	resolverCfg.ClientConfig = getClientConfig(cfg)
+	if resolverCfg.Registration, err = getRegistrationConfig(cfg, props); err != nil {
+		return resolverCfg, err
+	}
+	resolverCfg.QueryOptions = getQueryOptionsConfig(cfg)
+
+	// if registering, set DaprPort in meta, needed for resolution
+	if resolverCfg.Registration != nil {
+		if resolverCfg.Registration.Meta == nil {
+			resolverCfg.Registration.Meta = map[string]string{}
+		}
+
+		resolverCfg.Registration.Meta[resolverCfg.DaprPortMetaKey] = daprPort
+	}
+
+	return resolverCfg, nil
+}
+
+func getClientConfig(cfg configSpec) *consul.Config {
+	// If no client config use library defaults
+	if cfg.ClientConfig != nil {
+		return cfg.ClientConfig
+	}
+
+	return consul.DefaultConfig()
+}
+
+func getRegistrationConfig(cfg configSpec, props map[string]string) (*consul.AgentServiceRegistration, error) {
+	// if advanced registration configured ignore other registration related configs
+	if cfg.AdvancedRegistration != nil {
+		return cfg.AdvancedRegistration, nil
+	} else if !cfg.SelfRegister {
+		return nil, nil
+	}
+
+	var appID string
+	var appPort string
+	var host string
+	var httpPort string
+	var ok bool
+
+	if appID, ok = props[nr.AppID]; !ok {
+		return nil, fmt.Errorf("nr metadata property missing: %s", nr.AppID)
```

Comment:
`nr` in the error message may not make sense to the user. Could be `consul name resolution: ` which is the pattern I've seen in other components but there is not 100% consistency.

Reply:
These errors are actually all raised in the context of the consul name resolution component, so perhaps we should remove nr all together and leave it as

`return nil, fmt.Errorf("metadata property missing: %s", nr.AppID)`

then in the log it will look as so like this

`msg="failed to initialize name resolution resolver consul: metadata property missing:...`

instead of this

`msg="failed to initialize name resolution resolver consul: consul name resolution: metadata property missing:...`

what do you think?

Reply:
done - made the change let me know if you think otherwise